keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,773 @@
+
+### Goals
+
+#### Phase 1
+
+- Add an inline pod definition for issuing commands to containers.
+- Add an API object to send a request to trigger those commands.
+- Implement controller logic in a single *trusted* controller.
+
+#### Phase 2
+
+- Add support to send signals to containers.
+- Move controller logic into Kubelet.
+
+#### Phase 3
+
+- Add a probe to verify the results from the commands/signals if needed.
+
+### Non-Goals
+
+- Writing scripts, i.e., quiesce scripts, that are triggered by the commands are the responsibilities of users who set up pod definitions.
+
+## Proposal
+
+Define a new feature gate: `ContainerNotifier`.
+
+In phase 1, this API proposal adds an inline pod definition for sending commands to containers and an API object to send a request to trigger those commands.
+
+An external controller such as the application snapshot controller would signal when a command needs to run by creating a request notification API object. A SINGLE *trusted* controller (container notifier controller)  will be implemented to watch the objects and run the command when it is created and update its status accordingly.
+
+In phase 2, this API proposal adds an inline pod definition for container signals and allows the API object to send a request to trigger those signals. Logic in the container notifier controller will be moved to Kubelet. Kubelet watches the objects and runs the command when it is created and updates its status accordingly.
+
+In phase 3, a Probe may be added if needed as an inline pod definition to verify whether the signal is delivered or whether the command is run and results in the desired outcome.
+
+### API Changes
+
+#### Phase 1
+
+There are two parts in the API changes:
+* Add an inline pod definition for ContainerNotifierAction - define a command to run in the container.
+* Add a Notification API object - this is the request to trigger a container notifier.
+
+#### Phase 2
+
+* Add a signal to the inline pod definition:
+  * In the inline pod definition for ContainerNotifierAction, allow a signal to be defined and sent to the container.
+
+#### Phase 3
+
+* If needed, add an inline pod definition for Probe - probes the container to verify that ContainerNotifierAction has brought the container to the desired state.
+
+#### Phase 1 API Changes
+
+##### Inline Pod Definition for ContainerNotifier
+
+Add []ContainerNotifier to the Container struct:
+
+```
+type Container struct {
+     ......
+     // +optional
+     Lifecycle *Lifecycle `json:"lifecycle,omitempty" protobuf:"bytes,12,opt,name=lifecycle"`
+
+    // Notifiers can be triggered by Notification resources. Each notifier must
+    // have a unique name within the container definition, but the same name may be
+    // used in different containers of the same pod. When a Notification is created
+    // which selects a pod and matches the Notifier name, the Notifier will run in the
+    // container.
+    // +optional
+    Notifiers []ContainerNotifier
```

Comment:
do you think we could have named the lifecycle hooks and reused them ?

Reply:
Do we need this at pod level, to mean it applies to all containers ?

Reply:
I don't think reusing hook definitions is worth the effort - early versions of this had a new resource just for that and I found it quite cumbersome.  It's part of the pod, just like a probe.

Reply:
Re pod-level - I think that's an optimization (for laziness) which we can explore if it comes up.  I doubt it.